[WIP] [DONT MERGE] Enable demangling of C++ symbols in the stacktrace

ErnyTech opened the pull request:
This allows you to view demangled and pretty printed C++ symbols
in the stacktrace.
Is especially useful considering that D is getting more and more support for C++.

Now the C++ symbols in the stacktrace will be printed like this:
```
        test.d:7 [C++] Test<int>::SomeName(int, long, int) [0x55711efee36f]
        test.d:18 [C++] testcpp() [0x557228e17978]
```

This need the PR #3002

Geod24 commented:
FYI we now have https://github.com/dlang/dmd/pull/10964 in DMD.
We might be interested in having it here, too.

rainers commented:
FYI there was a similar PR: https://github.com/dlang/druntime/pull/2083

I think this one has better separation. For Windows, the VC mangling should be supported via UnDecorateSymbolName. Please also add a test case.

rainers commented on src/core/sys/windows/stacktrace.d:
```diff
@@ -294,7 +294,19 @@ private:
-        res ~= demangle(tempSymName, demangleBuf);
+        auto demangledName = demangle(tempSymName, demangleBuf);
+
+        version (Shared)
```
Why limit this to a shared build of the runtime?

ErnyTech replied:
you're right, for Windows it is not necessary

rainers replied:
Why is it necessary on other platforms. "Shared" is not a version for dlopen support, but whether the runtime itself is a shared library.

ErnyTech replied:
druntime must be launched with -ldl, otherwise the linker will not find dlopen.
I don't know how i can check this in a better way

rainers replied:
The compiler already adds -ldl to the linker command line (at least for linux):
https://github.com/dlang/dmd/blob/e8f984995598543279c07b9800d19a9e19321562/src/dmd/link.d#L756

ErnyTech replied:
Yes, but only if phobos is linked

ErnyTech replied:
Ok but if I create the object files with dmd and then link it manually? I'm not sure it's a good idea to depend on libdl for druntime

rainers replied:
Ah, I see that `rt_loadLibrary` is only implemented if `version(Shared)`, so it's consistent to not use it here, too. That restriction seems a bit strange, though.

ErnyTech replied:
We could do better if dmd supports weak symbols.
It would be enough to create a dlopen function that returns null marked as weak symbol, so if the user links libdl the real dlopen would be used

Geod24 commented on changelog/cpptrace.dd:
```diff
@@ -0,0 +1,51 @@
+Added experimental `C++` symbol demanling in the stacktrace
+
+This feature is availableby  passing the following switches to your executable (not to the compiler):
```
Ideally, the release note should start by describing the feature itself. You've been working on it for quite some time, so you know exactly what it entails, but not everyone does. Something like:
```
When a stack trace was generated by the runtime (e.g. when calling `Throwable.toString`), C++ names were not previously demangled. Hence the following code:
---
<INSERT CODE>
---
Would lead to the following stack trace:
---
<INSERT MANGLED STACK TRACE>
---

Starting from this release, C++ names will be demangled whenever possible, hence the previous stack trace will now show as:
---
<INSERT DEMANGLED STACK TRACE>
---

This feature is still experimental and must be opted-in by <BLAH BLAH>.
Demangling is only possible when <BLAH BLAH>.
```

Note that this is just an example, but it illustrate a few points:
- Always describe the feature first, what it brings to the user: imagine giving an elevator pitch about it;
- Try to show the difference between the old behavior and the new one: remember users might not be familiar with the old behavior, especially if someone looks at this release note in a few year;
- Include a code example whenever possible (which you did already);
- Put restrictions / limitations at the end;

Geod24 commented on src/core/internal/cpptrace.d:
```diff
@@ -0,0 +1,119 @@
+
+private struct CPPTrace
+{
+    __gshared CPPDemangle instance;
```
Don't we already have a singleton in the other module?

Geod24 commented on src/core/internal/cpptrace.d:
```diff
@@ -0,0 +1,119 @@
+    __gshared CPPDemangle instance;
+    __gshared Config config;
+
+    static this()
```
A `static this()` is run on each thread constructions. You want a `shared static this` here, no ?

Geod24 reviewed (commented):
Added a few comments about the second commit

Commit:
Add cppdemangle in druntime

Having a C++ symbol demangler is very useful, with this it will
be possible to implement the C++ symbol demangle in the stacktrace.

The demangling is done in Posix platform by function __cxa_demangle
presents in Itanium C++ ABI (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#demangler).

Itanium C++ ABI is used practically in all modern C++ compilers on Posix,
however old compilers (like GCC < 3.0) used a different name mangling,
this is not a problem because DMD in fact only supports Itanium C++ ABI on Posix.

For Windows instead the implementation it is provided by the UnDecorateSymbolName
function present in the Debug Help Library.
(https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-undecoratesymbolname)

Signed-off-by: Ernesto Castellotti <mail@ernestocastellotti.it>

Commit:
Enable demangling of C++ symbols in the stacktrace

This allows you to view demangled and pretty printed C++ symbols
in the stacktrace.
Is especially useful considering that D is getting more and more support for C++.

Now the C++ symbols in the stacktrace will be printed like this:
	test.d:7 [C++] Test<int>::SomeName(int, long, int) [0x55711efee36f]
        test.d:18 [C++] testcpp() [0x557228e17978]

Signed-off-by: Ernesto Castellotti <mail@ernestocastellotti.it>

RazvanN7 commented:
@ErnyTech What's the status on this?
